### PR TITLE
Ci test randomness

### PR DESCRIPTION
**Pull request by DickJC123**
This PR proposes two new features to how CI uses random data for testing:

1. A 'with np_random_seed(NNN)' syntax to allow a single test to run with a set seed without forcing determinism on other tests in the same file, and
2. A set_np_random_seed() util function to set the seed randomly and output that setting to the nosetests log file.  This permits a failure seen in a log file to be reproduced exactly.

Numerous individual tests have been also made more robust.  Most often the failures were the result of the finite difference method gradient not matching the symbol's gradient at a point when the gradient was either large (e.g. cube-root) or discontinuous (e.g. max).  In other cases, tolerances needed to be added or increased.

The proposed new approach is demonstrated on a simple test file of three tests.  Assuming that the second test needs a set seed for robustness, the file might currently appear as:

```
def test_op1():
    <op1 test>

def test_op2():
    np.random.seed(1234)
    <op2 test>

def test_op3():
    <op3 test>
```
Even though test_op3() is OK with nondeterministic data, it will have only a single dataset because it is run after test_op2, which sets the seed.  Also, if test_op1() were to fail, there would be no way to reproduce the failure, except for running the test individually to produce a new and hopefully similar failure.

With the proposed approach, the test file becomes:

```
set_np_random_seed()

def test_op1():
    <op1 test>

def test_op2():
    with np_random_seed(1234):
        <op2 test>

def test_op3():
    <op3 test>
```

set_np_random_seed() is in the global region of the test file and will set the seed differently for each run.  The value set will appear in the nosetests output.  The body of the test_op2 test now appears after the 'with np_random_seed(1234):' statement.  This has the effect of setting the seed before the test runs, then reinstating the rng state after the test is over (failed or passed).  Thus test_op3 will run with non-deterministic data.  Finally, if say test_op1() fails, the seed is known so the failure can be reproduced.  Further use of np.random.get_state() and np.random.set_state() can result in a modified test file that can be invoked as 'nosetests --verbose -s test_file.py:test_op1'. 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] ~For user-facing API changes, API doc string has been updated.~
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Intersting edge cases to note here

### CODE REVIEW

**Comment by cjolivier01** [comment]
Does this change the random behavior of tests which don’t call one of your new functions?

**Comment by DickJC123** [comment]
Tests that are downstream from a test that previously set a seed will now see random data, where before they would not have.  I also took out a global seed(42) call in test_operator_gpu.py, so some of those tests will see randomness for the first time.  

It’s not too hard to make these tests robust. Takes me less than an hour per failure to understand the reason and correct it. With this PR, we’ll be able to reinvoke any failure and add instrumentation to understand it.

File: python/mxnet/test_utils.py [hunk 1]
```diff
@@ -830,6 +834,8 @@ def check_numeric_gradient(sym, location, aux_states=None, numeric_eps=1e-3, rto
     if ctx is None:
         ctx = default_context()
 
+    _rng = get_rng()
```
Comment (larroy) [inline python/mxnet/test_utils.py]: why the _ prefix? is not a member?

File: python/mxnet/test_utils.py [hunk 2]
```diff
@@ -1538,3 +1546,53 @@ def discard_stderr():
...
+    The expected use of this function is to set the seed globally before
+    a suite of tests and output the set value to help reproduce any failures
+    that are dependent on the random data.  To fix the seed for a single test
+    without modifying the randomness of subsequent tests in the same file,
+    use 'np_random_seed'.
+    """
+    if seed is None:
+        seed = np.random.randint(0, np.iinfo(np.uint32).max)
+    if ostream is not None:
+        ostream.write('Setting np.random seed to %s.\n' % seed)
```
Comment (larroy) [inline python/mxnet/test_utils.py]: Why not use logging? I think is better to writing to stdout or stderr in general.

**Comment by larroy** [comment]
LGTM, modulo the comments above.

**Comment by piiswrong** [comment]
Sorry for having overlooked this. @DickJC123 Is it still relevant?

**Pull request closed without merging**